### ISSUE TITLE
List registered tasks at startup

### ISSUE THREAD

**Issue by mlissner** [rq/rq#1068]
Something I miss about Celery is that when started, the logger would list all the tasks it knew about. It makes it a lot easier to line up the tasks you're trying to run with the tasks that the worker is aware of. 

This is a small thing, but is it possible to do something similar for rq?

**Comment by selwin** [comment]
Sure, but listing all the tasks in the queue could result in a huge log file if there are lots of queued tasks. I think simply stating that there are `X tasks in Y queue` would be a good compromise. What do you think?

**Comment by mlissner** [comment]
Sorry, I meant this in a different way (though I like your idea too). I meant it should list something like:

```
Registered tasks:
my_code.some_module.some_task1
my_code.some_module.some_task2
my_code.some_module.some_task3
my_code.some_module.some_task4
my_code.some_module.some_task5
```

That way, you know that RQ knows which tasks it can run and you know the path to use to call them.

**Comment by selwin** [comment]
Yeah, I think this is something that we can do, but we should limit the output to ~100 tasks or so. Feel free to open a PR for this :)

**Comment by lowercase00** [comment]
This mostly happens in Celery because of the task decorator (and the manual task registration). @selwin  
Suggestion to to close this, mostly cosmetic and low priority, maybe reopen in the future..

**Comment by lowercase00** [comment]
@mlissner as this is fairly old and stale, I'm closing this. Feel free to reopen if you think this still makes sense.

**Comment by mlissner** [comment]
I still like this feature from Celery, but I defer, as ever, to the wisdom of the maintainers.

**Comment by lowercase00** [comment]
No worries, I'll reopen it as a feature request and leave the decision to @selwin then.

From what I understand from Celery/RQ. The reason this is possible with Celery, is that Celery requires task registration at startup, so Celery's main class does have that information when the worker starts - it always knows what it will run. RQ doesn't really know about which tasks a worker can run, since it's totally dynamic. So I'm not sure if that would be possible. 

We could start to worker, and right at startup check for tasks that exist in the queue and log that, but we wouldn't be logging "registered" tasks, but rather the "existing" tasks.

**Comment by mlissner** [comment]
> we wouldn't be logging "registered" tasks, but rather the "existing" tasks.

If that's what's possible, it doesn't seem useful. It might even be anti-useful, since it'd give the wrong impression that those tasks were the ones that rq knew about (and not others).

With Celery, it's just nice to know that all the duct tape and bailing wire is working properly. Celery knows _these_ tasks exist, and if you enqueue them, they will _probably_ get run.

**Comment by lowercase00** [comment]
Totally, I understand.

I might be missing something, but when we run a worker, we are just running a loop that is dequeuing stuff, and performing a function, the tasks are not registered in anyway at the start. We possibly could have a hook `on_app_start` that actually register tasks... but even then, your worker is a different service than the one that's enqueueing stuff, so this would be very fragile: if you start your worker before starting your service you wouldn't now anything. You could also register tasks on the worker it self `worker.register_task(...)` and then `worker.work()` but if you're doing that, would the logs really matter?

I really don't see a clean way to implement this, probably the best bet is to invest in powerful monitoring (`rq-dashboard`, `django-rq`, `rq-monitor`, improve `rq info` etc etc), so I still think this is most likely a `wont-fix` with low priority (not that it wouldn't be nice).

@selwin let us know if you can see any other solution.

**Comment by selwin** [comment]
I’m willing to accept a PR for this, more statistics and information will improve RQ’s usability and maintanability. I personally wanted a feature that lists how many jobs of X types are currently queued so this is a step towards that. 

I’m willing to accept a PR for this :)

We can implement this using a sorted set of registered tasks, with the type of callable as key and score as the number of jobs enqueued for a certain type of jobs (when it’s implemented).

**Comment by lowercase00** [comment]
I think the question is what **is** a registered task, we don't really have this concept now, nor the mechanism to know that at startup. This could be implemented looking at the queue, but it would only work for enqueued/scheduled jobs.

**Comment by selwin** [comment]
So the idea is to have output similar to this:

* `send_email`: 5 jobs
* `crawl_url`: 10 jobs
* `path.to.generate_report`: 2 jobs

How do `types` get added to this list? When a job gets enqueued (it also increments the count) and decrements it when executed.
